### PR TITLE
Improve the HTML code viewer

### LINKED ISSUE #97: HTML source code shouldn't have blue background
Ideally it would look like this

Screen Shot 2019-06-18 at 12 29 38 PM

cc  @mtharrison

### LINKED ISSUE #98: HTML source code should be able to select lines based on anchor tags
I can select a single line of source code on github using the anchor tag: https://github.com/denoland/deno_std/blob/master/http/server.ts#L7

It would be nice to be able to do this on deno.land too: https://deno.land/std/http/server.ts#L7

### PR DESCRIPTION
**Pull request by j-f1**
Fixes #98 

- [x] Add line numbers
- [x] Support highlighting single lines: https://deno.land/std/http/server.ts#L10
- [x] Support highlighting a range of lines: https://deno.land/std/http/server.ts#L11-L16
- [x] Support <kbd>shift</kbd> + click to select a range of lines
- [x] Fix #97 
- [x] Support dark theme
- [x] (optional) Redirect `https://deno.land/std/http/server.ts:10` to `https://deno.land/std/http/server.ts#L10` to allow clicking on links in stack traces

### CODE REVIEW

**Comment by bartlomieju** [comment]
> (optional) Redirect https://deno.land/std/http/server.ts:10 to https://deno.land/std/http/server.ts#L10 to allow clicking on links in stack traces

Clever idea

**Comment by bartlomieju** [comment]
@j-f1 would you mind addressing #97 in this PR as well?

**Comment by j-f1** [comment]
#97 has been addressed @bartlomieju

**Comment by bartlomieju** [comment]
@j-f1 I guess you're running it locally to check, can you share script you're using to serve `app`?

**Comment by j-f1** [comment]
I’m using the one @mtharrison [wrote](https://gist.github.com/mtharrison/8ee409e56dc174d94b9d415d09c6a6b0).

File: src/app.js [hunk 1]
```diff
@@ -213,16 +214,139 @@ function renderPrettyCode(pathname, code) {
...
+
+          const onHashChange = event => {
+            for (const el of document.querySelectorAll(
+              ".numbered-line.highlighted"
+            )) {
+              el.classList.remove("highlighted", "last-highlight");
+            }
+
+            const hash = parseHash();
+            if (hash.type === "single") {
```
Comment (bartlomieju) [inline src/app.js]: `hash` may be null, then you get:
```
file_server.ts:357 Uncaught TypeError: Cannot read property 'type' of null
```

File: src/app.js [hunk 2]
```diff
@@ -213,16 +214,139 @@ function renderPrettyCode(pathname, code) {
...
+              for (let line = hash.start; line <= hash.end; line++) {
+                const el = document.getElementById("L" + line);
+                el.classList.add("highlighted");
+                if (line === hash.end) {
+                  el.classList.add("last-highlight");
+                }
+              }
+              document.getElementById("L" + hash.start).scrollIntoView();
+            }
+            document.body.scrollTop -= 19 * 3.5;
```
Comment (bartlomieju) [inline src/app.js]: Do we need to change scroll? I hate when page "jumps" after click

File: src/app.js [hunk 3]
```diff
@@ -130,6 +148,12 @@ exports.lambdaHandler = async (event, context, callback) => {
     return callback(null, request);
   }
 
+  // URLs from error messages, i.e. https://deno.land/std/http/server.ts:10:24
+  const lineAndColumnMatch = pathname.match(/:(\d+)(?::\d+)?$/);
+  if (lineAndColumnMatch) {
+    return callback(null, redirect(pathname.replace(/:(\d+)(?::\d+)?$/, '#L$1')))
```
Comment (bartlomieju) [inline src/app.js]: It handles only start line, eg.:
```
http://localhost:4000/std/http/file_server.ts:10:20
// becomes:
http://localhost:4000/std/http/file_server.ts#L10
```

File: src/app.js (hunk 1 above)
Comment (j-f1) [reply src/app.js]: Fixed ✅

File: src/app.js (hunk 3 above)
Comment (j-f1) [reply src/app.js]: That’s correct — the second number is a column number, which we don’t support highlighting; although that could be a task for the future (i.e. add a red squiggly underline up to the next `\b`).

File: src/app.js (hunk 2 above)
Comment (j-f1) [reply src/app.js]: This is the same behavior as GitHub, and I’m just modifying the scroll here so the highlighted line isn’t pinned directly to the top of the page. I wrote the code to support a no-JS environment at least somewhat, so if you have JS disabled, the page will still scroll to a single selected line on load.

**Comment by j-f1** [comment]
> Having background was a bit more readable IMHO... WDYT?

cc @ry — what do you think?

File: src/app.js (hunk 3 above)
Comment (bartlomieju) [reply src/app.js]: You're right, that makes sense

**Comment by j-f1** [comment]
Added one more feature ⬆️

**Comment by ry** [comment]
@j-f1 Massive! 

Highlighting doesn't work for me: http://localhost:4000/std/http/server.ts#L17-L19

The URL at the top is kinda obnoxious - since it's the same as in the URL bar. Can you remove it? It would be cool to have a link to the github repo instead (but in normal sized text).

**Comment by bartlomieju** [comment]
Two more things to consider for future:
- adding [test server](https://gist.github.com/mtharrison/8ee409e56dc174d94b9d415d09c6a6b0) to repo 
- "copy permalink" - if user is viewing master branch there should be option to copy URL with current commit hash

**Comment by j-f1** [comment]
> Highlighting doesn't work for me: http://localhost:4000/std/http/server.ts#L17-L19

Try this link instead: http://localhost:4000/std/http/server.ts#L17-19 (note how there’s no second `L`)

> The URL at the top is kinda obnoxious - since it's the same as in the URL bar. Can you remove it? It would be cool to have a link to the github repo instead (but in normal sized text).

Should I format it as a bare URL or like this?

[View on GitHub](https://github.com/denoland/deno)

**Comment by ry** [comment]
> Should I format it as a bare URL or like this?

Sure

> Having background was a bit more readable IMHO... WDYT?

I prefer no background - it's less noise. I would love for this to be in a very similar style to the manual and other pages... minimal text.

File: src/app.js (hunk 3 above)
Comment (j-f1) [reply src/app.js]: ⏩ #102

**Comment by ry** [comment]
The clickable import links is everything I ever wanted. Thank you.


(Now we just need to get the types clickable :) )

**Comment by ry** [comment]
dark_light

the dark/light thing is neat

**Comment by j-f1** [comment]
This is good to merge from my end 🚀

**Comment by ry** [comment]
I'm getting this error from aws when trying to deploy...
```
Unable to import module 'app': Error
at Function.Module._load (module.js:474:25)
at Module.require (module.js:596:17)
at require (internal/module.js:11:18)
at Object.<anonymous> (/var/task/app.js:7:21)
at Module._compile (module.js:652:30)
at Object.Module._extensions..js (module.js:663:10)
at Module.load (module.js:565:32)
at tryModuleLoad (module.js:505:12)
at Function.Module._load (module.js:497:3)
```
investigating..

**Comment by j-f1** [comment]
Here’s a draft commit message:

```markdown
Improve the HTML code viewer

- Allow clicking on imported files to navigate to them
- Allow clicking on links written in the code
- Add line numbers
- Support linking to single lines and ranges of lines
- Follow the system dark mode setting where supported
- Redirect URLs produced by V8 error messages (e.g. `foo.ts:12:34`)
  to the appropriate line reference
- Remove the duplicated URL at the top of the page and replace it
  with a link to the GitHub repo of the current file

**Comment by j-f1** [comment]
Do you have to re-run `npm install`? Assuming `app.js` isn’t transformed by the build process, line 7 is the import of `linkify-urls`, which would fail if it isn’t installe.d

**Comment by ry** [comment]
Ok all fixed - https://deno.land/std/http/server.ts#L18-L24

It was using node 8 and also I hadn't run npm install.

**Pull request merged**